### PR TITLE
Paths and file management

### LINKED ISSUE #43: Improve directory and file management
Just noticed the existence of Pathlib, which seems nice, and worth switching to. Also need to improve some file management stuff so we can make a slightly more integrated PR out of this.
- [x] Use Pathlib for directory management
  - Now that I'm using this more, it creates a nice point where we can separate out all the paths from the monolithic `DictConfig` and start passing them around individually to subroutines. So I am now endeavouring to do a bigger push with this.
- [x] Save data from analyses so they may be plotted outside of a simulation
- [x] Fix some bugs/refine directory structure

### PR DESCRIPTION
**Pull request by alex404**
So this does quite a fair bit of refactoring.

- I setup the analyses in `statistics.py` to return dataclass instances to make things more manageable and readable. 
- I also made a push to a more pathlib style of directory management, in particular in the runner submodules.
- Generally going for a style where the at the top of the framework command (e.g. intialize, analyze or train) all necessary variables are extracted out of the DictConfig.
- Saving analysis results to disk was the original push for this, and that's done too.

I definitely wasn't perfectly complete or consistent with how I did this, but I think things are significantly better organized than they were before. Ideally this provides a good reference for how to factor/refactor such code in other parts of the library. Definitely open to feedback on review.x

Resolves #43

### CODE REVIEW

**Comment by alex404** [comment]
This probably seems more complicated than it is. The bulk of the changes are moving blocks of arguments around into dataclasses.

**Comment by alex404** [comment]
Actually hold on just discovered a bug

**Comment by alex404** [comment]
Okay seems to be fine now

**Review by fabioseel — COMMENTED** [review]
Seems mostly fine, just want to hear your comment thoughts on my comment in plot (moving the inner function out to module scope)
and in statistics (using the defined dataclasses already within the functions)

File: retinal_rl/analysis/plot.py [hunk 1]
```diff
@@ -48,33 +41,53 @@ def plot_transforms(
...
     )
 
-    # Create a figure with subplots for each transform
     fig, axs = plt.subplots(num_transforms, 1, figsize=(20, 5 * num_transforms))
     if num_transforms == 1:
         axs = [axs]
 
     transform_index = 0
 
+    def make_image_grid(arrays: List[FloatArray], nrow: int) -> FloatArray:
```
[the remark this answers is not in the document]
Comment (alex404) [reply retinal_rl/analysis/plot.py:50]: Not sure. There's a pytorch function that does this, and this is what Claude spat out when I asked it for a version that would work with numpy arrays. I can move out it though. Probably makes more sense.

File: retinal_rl/analysis/statistics.py [hunk 2]
```diff
@@ -34,7 +102,7 @@ def transform_base_images(
         src, _ = base_dataset[np.random.randint(base_len)]
         images.append(src)
 
-    results: Dict[str, Dict[str, Dict[float, List[Tensor]]]] = {
+    results: Dict[str, Dict[str, Dict[float, List[FloatArray]]]] = {
```
[the remark this answers is not in the document]
Comment (alex404) [reply retinal_rl/analysis/statistics.py]: Eh, not sure if changing things really helped but I did it as per request ;)

**Review by fabioseel — CHANGES_REQUESTED** [review]
just a minor change to avoid the duplicate code

File: retinal_rl/analysis/statistics.py [hunk 3]
```diff
@@ -34,29 +102,33 @@ def transform_base_images(
...
+    )
 
-    transforms: List[Tuple[str, nn.Module]] = []
-    transforms += [
-        ("source_transforms", transform) for transform in imageset.source_transforms
-    ]
-    transforms += [
-        ("noise_transforms", transform) for transform in imageset.noise_transforms
-    ]
+    for transform in imageset.source_transforms:
```
[the remark this answers is not in the document]
Comment (alex404) [reply retinal_rl/analysis/statistics.py]: As you wish.

**Review by fabioseel — APPROVED** [review]
Looks good to me now :)

**Pull request merged**